lib/custom-resources/endpoint-ips.ts

```diff
@@ -0,0 +1,69 @@
+import { EC2, NetworkInterface } from "@aws-sdk/client-ec2";
+import type { OnEventRequest, OnEventResponse } from "aws-cdk-lib/custom-resources/lib/provider-framework/types";
+
+const ec2 = new EC2({ useFipsEndpoint: true });
+
+/**
+ * Handle CloudFormation Custom Resource events using the CDK Custom Resource Provider Framework.
+ *
+ * This will, given a VPC Endpoint ID, determine the full list of IP addresses associated with
+ * that endpoint and return data in a list suitable for an Elastic Load Balancing V2 Target
+ * configuration. The output can be passed directly (via a `GetAtt`) to the underlying
+ * CloudFormation resource.
+ */
+export async function onEvent(event: OnEventRequest): Promise<OnEventResponse> {
+  // There is no data to provide on a Delete operation and no resources were
+  // actually created so there's nothing for us to actually destroy.
+  if (event.RequestType === "Delete") {
+    return {};
+  }
+
+  const endpointId = event.ResourceProperties.VpcEndpointId;
+  if (!endpointId) {
+    throw new Error("VpcEndpointId property is required");
+  }
+  // Every (or at least nearly every) VPC endpoint uses port 443, so this is a really
+  // safe default but we can use a different value if one is provided.
+  const port = event.ResourceProperties.Port ?? 443;
+
+  const endpointInterfaces = await getEnisForVpcEndpoint(endpointId);
+  if (!endpointInterfaces.length) {
+    throw new Error(`Vpc Endpoint "${endpointId}" does not exist (or does not have ENIs)`);
+  }
+
+  // On either a Create or Update event, we need to re-determine the IP addresses
+  // as they may change if the resource changes.
+  return {
+    PhysicalResourceId: endpointId,
+    Data: {
+      Targets: endpointInterfaces.map((eni) => ({
+        Port: port,
+        Id: eni.PrivateIpAddress,
+        AvailabilityZone: eni.AvailabilityZone,
+      })),
+    },
+  };
+}
+
+async function getEnisForVpcEndpoint(vpcEndpointId: string): Promise<NetworkInterface[]> {
+  const interfaceIds = (
+    await ec2.describeVpcEndpoints({
+      Filters: [
+        {
+          Name: "vpc-endpoint-id",
+          Values: [vpcEndpointId],
+        },
+      ],
+    })
+  )?.VpcEndpoints?.flatMap((endpoint) => endpoint.NetworkInterfaceIds).filter((eni): eni is string => !!eni);
```

Comment:
Is this a little trick to name anonymous functions?

Reply:
Not sure which part of this you're talking about but to expand this line out:

```ts
VpcEndpoints // Get the VPC endpoints
  // If VpcEndpoints exists, flatMap it to a list of NetworkInterfaceIds.
  // Because we are going to make a call for all of these, it's easier to have
  // a string[] than a string[][] at the end of this
  ?.flatMap((endpoint) => endpoint.NetworkInterfaceIds)
  // Only keep things that are truthy -- in some cases, `NetworkInterfaceIds` may
  // be undefined or something may happen that something else gets in there. But
  // prior to this filter, TSC is worried that our type overall is
  // `(string | undefined)[]`; so we want to filter out any undefined items.
  // Typically, we could just do `filter((eni) => eni)` which would only keep
  // truthy things but TypeScript isn't "smart" enough to understand that everything
  // left would be a string. So we have to instead specifically use a TypeScript
  // User-Defined Type Guard type for the function
  // (https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards)
  // This also means that we have to use `!!eni` instead of just `eni` because we
  // have to return a boolean to appease tsc (where as eni isn't a boolean, it's
  // string | undefined).
  .filter((eni): eni is string => !!eni);
```

Reply:
That's not actually my question but I appreciate the answer as I continue to learn at least one new thing about TypeScript in your PRs. I was referring to the `eni is string` portion instead.

Reply:
Oh yeah, So `eni is string` is a "return type" for a type predicate. It says, "this function returns a boolean and if true, this is a string; if false, it's something else". https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

Which is also why we need the `!!eni` coercion -- to force this to a `boolean`. Because TypeScript gets mad if your type predicate relies on truthiness instead of just being `true`/`false`.